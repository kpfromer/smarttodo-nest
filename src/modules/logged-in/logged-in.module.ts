import { Module } from '@nestjs/common';
import { LoggedInService } from './logged-in.service';

const servicesViaProviders = [
  { provide: LoggedInService, useValue: LoggedInService }
];

@Module({
  providers: servicesViaProviders,
  exports: servicesViaProviders
})
export class LoggedInModule {}
