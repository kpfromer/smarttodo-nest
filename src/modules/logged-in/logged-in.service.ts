import { Injectable } from '@nestjs/common';
import { ModelType } from 'typegoose';
import { Types } from 'mongoose';
import ObjectId = Types.ObjectId;
import { Document } from 'mongoose';

export type WithMongoId<T> = T & {
  _id: ObjectId
}

export type WithPlainId<T> = T & {
  id: string
}

export type CreatedLoggedInModel<T> = T & Document & {
  userId: string;
}

@Injectable()
export class LoggedInService<T, R = WithMongoId<T>> {

  constructor (protected readonly model: ModelType<T>) {}

  sanitize(document: CreatedLoggedInModel<T>) {
    // Convert to regular object
    const plainDocument = document.toObject();
    // Convert _id's ObjectId type to just a string
    const stringId = plainDocument._id.toHexString();
    delete plainDocument.userId;
    delete plainDocument._id
    return {
      // @ts-ignore
      ...plainDocument,
      id: stringId
    }
  }

  async getAll(userId: string, conditions = {}): Promise<WithPlainId<T>[] | null> {
    return (await this.model.find({ ...conditions, userId }).exec() as CreatedLoggedInModel<T>[]).map(this.sanitize);
  }

  async getById(userId: string, id: string): Promise<WithPlainId<T> | null> {
    return this.sanitize(await this.model.findOne({ userId, _id: id }).exec() as CreatedLoggedInModel<T>);
  }

  async create(userId: string, item: T) {
    // @ts-ignore
    return this.sanitize(await this.model.create({ ...item, userId }) as CreatedLoggedInModel<T>);
  }

  async updateById(userId: string, id: string, newItem: T & { id?: string }) {
    delete newItem.id;
    return this.sanitize(await this.model.updateOne({ userId, _id: id }, newItem).exec());
  }

  async deleteById(userId: string, id: string) {
    return await this.model.deleteOne({ userId, _id: id }).exec();
  }
}