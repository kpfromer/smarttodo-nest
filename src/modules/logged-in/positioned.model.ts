import { prop, Typegoose } from 'typegoose';

export abstract class PositionedModel extends Typegoose {
  @prop({ required: true })
  readonly weight: number;
}