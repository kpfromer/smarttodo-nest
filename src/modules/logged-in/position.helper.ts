import { ModelType } from 'typegoose';
import { WithMongoId } from './logged-in.service';
import { Config } from '../../config';

export class PositionHelper<T> {

  constructor(private readonly model: ModelType<T>) {}

  async getItemsWithWeight<T>(conditions?: object): Promise<(WithMongoId<T> & { weight: number })[] | null> {
    return await this.model.find(conditions)
      .sort({
        weight: 1
      })
      .lean()
      .exec();
  };

  async reassignWeightByIndex<T>(items: WithMongoId<T>[]) {
    const bulkUpdate = items.map((item, position) => {
      const updatedWeightItem = {
        // @ts-ignore (see https://github.com/Microsoft/TypeScript/pull/13288)
        ...item,
        weight: position * Config.get('/todoWeightStep')
      };

      return {
        updateOne: {
          filter: { _id: item._id },
          update: updatedWeightItem
        }
      }
    });

    await this.model.bulkWrite(bulkUpdate);
  };
// TODO: FIX PARAM
  async recheckIsInPosition<T>(items: WithMongoId<T>[], item: WithMongoId<T>, position: number) {
    // Recheck if the item is that the request position
    // if not, the weights of the items are no longer stable and need to be reassigned
    const id = item._id.toHexString();
    if (items[position]._id.toHexString() !== id) {
      const realIdPosition = items.findIndex(item => item._id.toHexString() === id);
      if (realIdPosition === -1) {
        throw new Error(`Item with id: ${id} was removed!`);
      }
      const repositionedItems = [
        ...items.slice(0, realIdPosition),
        ...items.slice(realIdPosition + 1)
      ];
      // Re add item back into correct location
      repositionedItems.splice(position, 0, item);
      await this.reassignWeightByIndex(repositionedItems);
    }
  };
  // TODO: FIX PARAM
  async getWeight<T>(conditions: object = {}, position: number = null, oldPosition = null, operation: (weight: number) => Promise<{ item: WithMongoId<T>, returnValue: any }>) {
    const items = await this.getItemsWithWeight<T>(conditions);

    if (position === null)
      position = items.length - 1;

    let diff = -1;
    // Fixes comparison of same item weight with other causing the weight to change but will still be in same position
    if (oldPosition !== null)
      diff = position - oldPosition === 1 ? 0 : -1;

    let weight;

    if (position === 0 && typeof items[0] !== 'undefined') {
      weight = items[0].weight - Config.get('/todoWeightStep');
    } else if (position === items.length - 1 && typeof items[items.length - 1] !== 'undefined') {
      weight = items[items.length - 1].weight + Config.get('/todoWeightStep');
    } else if (typeof items[position + diff] !== 'undefined' && typeof items[position + diff + 1] !== 'undefined') {
      weight = (items[position + diff].weight + items[position + diff + 1].weight) / 2;
    } else {
      weight = position * Config.get('/todoWeightStep');
    }

    const { item, returnValue } = await operation(weight);

    // Ignoring the promise so that the response is faster
    this.recheckIsInPosition<T>(items, item, position);

    return returnValue;
  };
}
