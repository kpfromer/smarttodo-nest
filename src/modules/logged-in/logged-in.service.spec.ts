import { LoggedInService } from './logged-in.service';

describe('LoggedInService', () => {
  let loggedInService, mockModel, modelInstance, userId;

  beforeEach(() => {
    mockModel = {
      find: jest.fn(),
      findOne: jest.fn(),
      create: jest.fn(),
      findOneAndUpdate: jest.fn(),
      deleteOne: jest.fn()
    };
    modelInstance = jest.fn();
    userId = 'a-bunch-of-numbers-letters';

    // jest.spyOn(Config, 'get')
    //   .mockImplementation(value => value === '/userIdProperty' ? userIdProperty : undefined);
    loggedInService = new LoggedInService(mockModel);
  });

  describe('find', () => {
    beforeEach(() => {
      mockModel.find.mockImplementation(() => ({
        lean: () => ({
          exec: () => Promise.resolve(modelInstance)
        })
      }));
    });

    it('finds an instance using the userId but removes the property on return', async () => {
      const result = await loggedInService.find(userId);

      expect(result).toBe(modelInstance);
      expect(mockModel.find).toHaveBeenCalledWith({ userId }, '-userId');
    });
    it('finds an instance with a condition', async () => {
      const conditions = {
        hello: 'world'
      };
      await loggedInService.find(userId, conditions);

      expect(mockModel.find.mock.calls[0][0]).toEqual({
        ...conditions,
        userId
      });
    });
  });
  describe('getAll', () => {
    beforeEach(() => {
      jest.spyOn(loggedInService, 'find').mockResolvedValue(modelInstance);
    });
    it('finds all', async () => {
      const result = await loggedInService.getAll(userId);
      expect(result).toBe(modelInstance);
      expect(loggedInService.find).toHaveBeenCalledWith(userId);
    });
  });
  describe('getById', () => {
    beforeEach(() => {
      mockModel.findOne.mockReturnValue({
        lean: () => ({
          exec: () => Promise.resolve(modelInstance)
        })
      });
    });

    it('finds one item by it\'s id', async () => {
      const id = 'fadsfad1r';
      const result = await loggedInService.getById(userId, id);
      expect(result).toBe(modelInstance);
      expect(mockModel.findOne).toHaveBeenCalledWith({ userId, _id: id });
    });
  });
  describe('create', () => {
    beforeEach(() => {
      mockModel.create.mockResolvedValue(modelInstance);
    });

    // TODO: Update when nestjs allows for proper validation of multiple items
    // it('creates multiple items', async () => {
    //   const items = [
    //     {
    //       name: 'jake',
    //       age: 29
    //     },
    //     {
    //       name: 'paul',
    //       age: 54
    //     },
    //     {
    //       name: 'mark',
    //       age: 34
    //     }
    //   ];
    //
    //   const result = await loggedInService.create(userId, items);
    //   expect(result).toBe(modelInstance);
    //   expect(mockModel.create).toHaveBeenCalledWith([
    //     {
    //       userId,
    //       name: 'jake',
    //       age: 29
    //     },
    //     {
    //       userId,
    //       name: 'paul',
    //       age: 54
    //     },
    //     {
    //       userId,
    //       name: 'mark',
    //       age: 34
    //     }
    //   ]);
    // });
  });
  describe('updateById', () => {
    beforeEach(() => {
      mockModel.findOneAndUpdate.mockReturnValue({
        exec: () => Promise.resolve(modelInstance)
      });
    });

    it('updates an item by it\'s id', async () => {
      const id = 'afadsfasd';
      const newItem = {
        _id: 'id',
        name: 'different',
        age: 6246234
      };
      const result = await loggedInService.updateById(userId, id, newItem);
      expect(result).toBe(modelInstance);
      expect(mockModel.findOneAndUpdate).toHaveBeenCalledWith({ userId, _id: id }, { ...newItem, userId });
    });
  });
  describe('deleteById', () => {
    beforeEach(() => {
      mockModel.deleteOne.mockReturnValue({
        exec: () => Promise.resolve(modelInstance)
      });
    });

    it('removes one item by id', async () => {
      const id = 'adfasdfasd';
      await loggedInService.deleteById(userId, id);
      expect(mockModel.deleteOne).toHaveBeenCalledWith({ userId, _id: id });
    });
  });
});