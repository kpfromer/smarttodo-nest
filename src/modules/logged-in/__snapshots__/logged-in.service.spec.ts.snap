// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoggedInService create creates an item with user id 1`] = `
Object {
  "id": "312",
  "name": "Macy",
}
`;

exports[`LoggedInService getAll returns sanitized data by user id 1`] = `
Array [
  Object {
    "id": "1",
    "name": "mick",
  },
  Object {
    "id": "2",
    "name": "joe",
  },
]
`;

exports[`LoggedInService getById gets all items by user id and item id 1`] = `
Object {
  "id": "666",
  "name": "jack daniels",
}
`;

exports[`LoggedInService updateById updates an item by id 1`] = `
Object {
  "id": "update_id",
  "name": "jack",
}
`;
